lms/models/lti_registration.py

```diff
@@ -37,3 +39,62 @@ class LTIRegistration(CreatedUpdatedMixin, BASE):
     application_instances = sa.orm.relationship(
         "ApplicationInstance", back_populates="lti_registration"
     )
+
+    @staticmethod
+    def urls(issuer, client_id):
+        auth_login_url = key_set_url = token_url = None
+        issuer_host = urlparse(issuer).netloc
+
+        if issuer == "https://blackboard.com":
+            # For blackboard all LTI1.3 apps use the same registration
+            auth_login_url = "https://developer.blackboard.com/api/v1/gateway/oidcauth"
+            key_set_url = f"https://developer.blackboard.com/api/v1/management/applications/{client_id}/jwks.json"
+            token_url = (
+                "https://developer.blackboard.com/api/v1/gateway/oauth2/jwttoken"
+            )
+        elif issuer.endswith(".instructure.com"):
+            # Different instructure hosted canvas use the same URLs
+            auth_login_url = "https://canvas.instructure.com/api/lti/authorize_redirect"
+            key_set_url = "https://canvas.instructure.com/api/lti/security/jwks"
+            token_url = "https://canvas.instructure.com/login/oauth2/token"
+        elif issuer.endswith(".brightspace.com"):
+            # Hosted D2L use the issuer URL as the base
+            auth_login_url = (
+                urlparse("https://brightspace.com/d2l/lti/authenticate")
+                ._replace(netloc=issuer_host)
+                .geturl()
+            )
+            key_set_url = (
+                urlparse("https://brightspace.com/d2l/.well-known/jwks")
+                ._replace(netloc=issuer_host)
+                .geturl()
+            )
+            token_url = (
+                urlparse("https://brightspace.com/core/connect/token")
+                ._replace(netloc=issuer_host)
+                .geturl()
+            )
+
+        elif issuer.endswith(".moodlecloud.com"):
+            # Hosted moodle use the issuer URL as the base
+            auth_login_url = (
+                urlparse("https://moodlecloud.com/mod/lti/auth.php")
+                ._replace(netloc=issuer_host)
+                .geturl()
+            )
+            key_set_url = (
+                urlparse("https://moodlecloud.com/mod/lti/certs.php")
+                ._replace(netloc=issuer_host)
+                .geturl()
+            )
+            token_url = (
+                urlparse("https://moodlecloud.com/mod/lti/certs.php")
+                ._replace(netloc=issuer_host)
+                .geturl()
+            )
+
+        return {
+            "auth_login_url": auth_login_url,
+            "key_set_url": key_set_url,
+            "token_url": token_url,
+        }
```

Comment:
I don't think the model is the place for this. This really seems like something that should be in the admin view as it's only used there, not in the normal run of things. 

In general I think we're moving away from adding code to the models, especially as here it's a static method which doesn't rely on any details for the instance.

Reply:
The amount of config in the code here makes me a bit uncomfortable. This might be a personal style thing, but I'd prefer a configuration that could be updated and some code which will then use that config.

I wanted to try this myself to make sure it's possible and came up with the following (just a suggestion):

```python
URL_SUGGESTIONS = {
    "https://blackboard.com": {
        "auth_login_url": "https://developer.blackboard.com/api/v1/gateway/oidcauth",
        "key_set_url": "https://developer.blackboard.com/api/v1/management/applications/{client_id}/jwks.json",
        "token_url": "https://developer.blackboard.com/api/v1/gateway/oauth2/jwttoken",
    },
    ".instructure.com": {
        "auth_login_url": "https://canvas.instructure.com/api/lti/authorize_redirect",
        "key_set_url": "https://canvas.instructure.com/api/lti/security/jwks",
        "token_url": "https://canvas.instructure.com/login/oauth2/token",
    },
    ".brightspace.com": {
        "auth_login_url": "https://{issuer_host}/d2l/lti/authenticate",
        "key_set_url": "https://{issuer_host}/d2l/.well-known/jwks",
        "token_url": "https://{issuer_host}/core/connect/token",
    },
    ".moodlecloud.com": {
        "auth_login_url": "https://{issuer_host}/mod/lti/auth.php",
        "key_set_url": "https://{issuer_host}/mod/lti/certs.php",
        "token_url": "https://{issuer_host}/mod/lti/certs.php",
    },
}

def urls(issuer, client_id):
    for suffix, url_templates in URL_SUGGESTIONS.items():
        if issuer.endswith(suffix):
            issuer_host = urlparse(issuer).netloc

            return {
                key: template.format(issuer_host=issuer_host, client_id=client_id)
                for key, template in url_templates.items()
            }

    return {"auth_login_url": None, "key_set_url": None, "token_url": None}
```

This config could then live in a JSON file if we wanted to keep the code slim.